RenderPassEditor : Add Rename Render Pass menu item

murraystevenson opened the pull request:
This adds a "Rename Selected Render Pass" menu item to the "Name" column's popup menu with the ability to rename a render pass within the edit scope it was originally created. This action covers simple cases, such as fixing a typo in a newly created render pass name.

References to the old render pass name outside of the originating edit scope are left unchanged by this action and would still need to be updated by the user, including those in other edit scopes.

![renameSelectedRenderPass](https://github.com/user-attachments/assets/29f90132-0694-463d-b55b-1ee0ae10d3d3)

tomc-cinesite commented:
@murraystevenson  - this looks like its really going to help people, thanks so much! This topic has come up a lot.

Q/unsolicited thought re:

> ... with the ability to rename a render pass within the edit scope it was originally created. 

Sorry not been able to test a build of this as on mac so please take with a pinch of salt - is this item only enabled if the target edit scope is the one that made the pass? When I watched the gif I think I misunderstood the warning message a little until I read the above description.

As it said "target edit scope" it made me think I could rename passes midway through the graph and/or it might only change the name within the scope itself. What do you think to something like:

> ⚠️  References to this pass by name will need to be manually updated

Can this be used in "source" mode too?

johnhaddon reviewed (commented):
Thanks Murray! Implementation LGTM but I've done a bit of nitpicking/hand-wringing about the messaging. Not sure my suggestions are better, but might be worth considering - see comments inline.

@tomc-cinesite, I think I can answer a couple of your questions :

> is this item only enabled if the target edit scope is the one that made the pass?

The operation only works if the target edit scope is the one that made the pass. But the menu item is enabled even the pass wasn't created in this scope - this lets us pop up a more informative warning/error instead of just greying out the menu item. The other option would be to grey out the icon but add a tooltip explaining the issue, but that seems less discoverable.

> Can this be used in "source" mode too?

No - source mode never makes edits inside an EditScope. Although we _could_ search for and edit the name in a RenderPasses node outside any edit scope, I'm not sure how useful that would be. And I think defining expectations around what else gets renamed automatically would be harder in source mode.

murraystevenson commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -434,6 +434,16 @@ def __columnContextMenuSignal( self, column, pathListing, menuDefinition ) :
 			# Render pass operations
 
+			menuDefinition.append(
+				"Rename Selected Render Pass",
```
It should indeed! Added in 8edbec8.

murraystevenson commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -489,6 +499,72 @@ def __addRenderPass( self, renderPass, editScope ) :
+			if renderPassEdits is not None :
+				for row in Gaffer.Spreadsheet.RowPlug.Range( renderPassEdits["edits"] ) :
+					if row["name"].getValue() == oldName :
+						row["name"].setValue( newName )
```
Added to EditScopeAlgo in fbd7cef and used in 7738f99.

After putting it in EditScopeAlgo, I did wonder if `renameRenderPass()` should also be able to rename RenderPassOptionEdits in an EditScope without a RenderPasses processor, but I've left that out for now...

murraystevenson commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -434,6 +434,16 @@ def __columnContextMenuSignal( self, column, pathListing, menuDefinition ) :
+			menuDefinition.append(
+				"Rename Selected Render Pass",
+				{
+					"command" : Gaffer.WeakMethod( self.__renameSelectedRenderPass ),
```
I was a bit hesitant to make it behave too much like the plug popups as it is doing more work than just editing the field. Maybe this is distinction is clear enough with the renaming happening via a Dialog rather than a Popup? 

If you want to give it a try, I've added the double-click and keypress handling in c5b9551. One challenge with renaming based on the keyPress event is returning focus back to the PathListingWidget after the Dialog is closed so you can further navigate with the keyboard. I've yet to find a reliable way of doing so...

johnhaddon replied:
Thanks for giving that a bash - I think I like it. The centred dialogue, additional messaging and the need to commit by pressing the button do make this feel like a "heavier" operation, to the point where it seems reasonable to expect it is doing more.

One other thing I would love is to have the "editscope blue" background on the render passes which were created in the current edit scope, for a quick visual indication of what can be renamed. But I think that's out of scope for this PR.

murraystevenson commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -489,6 +499,72 @@ def __addRenderPass( self, renderPass, editScope ) :
+			with GafferUI.PopupWindow() as self.__popup :
+				with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 ) :
+					GafferUI.Image( "warningSmall.png" )
+					GafferUI.Label( "<h4>Unable to rename render pass as it was not created in {}.</h4>".format( editScope.relativeName( self.scriptNode() ) ) )
```
I was mostly worried that something terse like "Pass was not created..." might be a little ambiguous as to whether it was describing the reason or the result (We couldn't rename because the pass wasn't created / We couldn't rename so the pass wasn't created). 

Having said that, I don't mind the two line split suggestion, so I've tried that in e3cc26c.

johnhaddon replied:
Thanks for giving that a go - I do think I prefer it.

murraystevenson commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -489,6 +499,72 @@ def __addRenderPass( self, renderPass, editScope ) :
+			title = "Rename Render Pass",
+			confirmLabel = "Rename",
+			defaultName = selectedRenderPasses[0],
+			warning = "The render pass will only be renamed inside the target edit scope."
```
I've reworked the Dialog and the messaging a bit in 2140bc5. 

![image](https://github.com/user-attachments/assets/89d25b74-ae38-41b1-9697-eeb079726228)

I do think it's an improvement but it still bothers me a little so if either of you have final tweaks I'll be happy to accommodate. 

I've avoided the use of "Dowstream" as while that makes sense for EditScopes, there's nothing stopping a user from having placed a `${renderPass}` dependant Spreadsheet or NameSwitch upstream of the current edit scope.

murraystevenson replied:
Thanks, I've made the adjustments you've suggested and think we should go with them. The switch to "info" does feel much friendlier. Changes in 42b68dc.

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -1492,3 +1492,35 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::renderPassesReadOnlyRe
+		throw IECore::Exception( fmt::format( "Render Pass \"{}\" already exists", newName ) );
+	}
+
+	if( std::find( renderPasses->readable().begin(), renderPasses->readable().end(), oldName ) != renderPasses->readable().end() )
+	{
+		auto renderPassesCopy = renderPasses->copy();
+		std::replace( renderPassesCopy->writable().begin(), renderPassesCopy->writable().end(), oldName, newName );
+		namesPlug->setValue( renderPassesCopy );
+	}
```
It seems a little odd that we're willing to ignore the absence of `oldName` here, but not to ignore the absence of `renderPassesProcessor`. Both cases are semantically the same - they mean the pass wasn't created here.

> I did wonder if `renameRenderPass()` should also be able to rename RenderPassOptionEdits in an EditScope without a RenderPasses processor

I think I like that - it would resolve the ambiguity above, and would allow someone to attempt the mind-reading "fix up my whole graph for me" operation.

murraystevenson replied:
> I think I like that - it would resolve the ambiguity above, and would allow someone to attempt the mind-reading "fix up my whole graph for me" operation.

Cool, someone attempting that style of fix ups was what came to mind there. I've addressed some of the ambiguity, supported renaming RenderPassOptionEdits without a RenderPasses processor and added the return true/false suggestion below in 8edae92.

Last bit of ambiguity is whether we should also be throwing if there's a collision with `newName` already existing in RenderPassOptionEdits...

johnhaddon replied:
> Last bit of ambiguity is whether we should also be throwing if there's a collision with newName already existing in RenderPassOptionEdits...

Oof. I think I would say that a single-row-per-pass is a guarantee we want the API to provide, so we should be throwing. We'd need to check before updating the RenderPasses node as well, so there aren't any side-effects when we do throw.

But what does that mean for the UI?

murraystevenson replied:
Updated to throw when `newName` already exists in RenderPassOptionEdits and added `renameRenderPassNonEditableReason` (bit of a mouthful there) to return the reason in f88d94e.

> But what does that mean for the UI?

d2b2360 makes the UI display the reason as a popup after you try and rename to a colliding name. I think it would be better to prevent the user from clicking `Rename` in the dialogue like what happens when they enter a render pass name that collides with the names in the `renderPass:names` option, but the post-popup was a bit more straightforward than attempting to stick an additional check into a dialogue that's being shared with the add render pass functionality...

johnhaddon commented on include/GafferScene/EditScopeAlgo.h:
```diff
@@ -162,6 +162,8 @@ GAFFERSCENE_API const Gaffer::GraphComponent *renderPassOptionEditReadOnlyReason
 
 GAFFERSCENE_API const Gaffer::GraphComponent *renderPassesReadOnlyReason( const Gaffer::EditScope *scope );
 
+GAFFERSCENE_API void renameRenderPass( Gaffer::EditScope *scope, const std::string &oldName, const std::string &newName );
```
I wonder if there's any value in returns true/false to indicate whether anything happened or not? I'm thinking about someone using this to attempt renaming throughout a graph - might be useful to be able to give a summary of what was affected afterwards?

murraystevenson replied:
Great, I like the idea. As mentioned above it's part of 8edae92.

johnhaddon reviewed (commented):
Thanks for the update Murray - definitely feeling like worthwhile improvements. I made a few take-it-or-leave-it comments inline, but I think we're pretty much there.

This has got me thinking about the slightly more ambitious alternative though, which would use a new RenameRenderPasses (or ShuffleRenderPasses?) node within the EditScope to allow local renaming of passes created upstream. I suspect we've talked about it before and I've forgotten what our objections were. But right now I feel like it might be quite nice in conjunction with blue-background-styling on the cell, and an inspector-style history to allow introspection. Definitely not proposing we do that right now, but would be curious to know your thoughts...

murraystevenson commented:
Thanks for the input, I do prefer where we've ended up! I've squashed in the last round of fixups and rebased. Last few changes are commented inline.

> This has got me thinking about the slightly more ambitious alternative though, which would use a new RenameRenderPasses (or ShuffleRenderPasses?) node within the EditScope to allow local renaming of passes created upstream. I suspect we've talked about it before and I've forgotten what our objections were. But right now I feel like it might be quite nice in conjunction with blue-background-styling on the cell, and an inspector-style history to allow introspection. Definitely not proposing we do that right now, but would be curious to know your thoughts...

I do think there's a lot of value in a ShuffleRenderPasses node for more procedural approaches. I think the action here is more for the "fix up my typo" style renaming where you don't really want to preserve the history. A separate "Shuffle to..." action in the Render Pass Editor might distinguish the more procedural approach you're suggesting along with the blue-backgrounds and history to support it...

johnhaddon commented:
Thanks Murray - updates LGTM. I suggest we resolve the to-throw-or-not-to-throw question in today's catch up.

johnhaddon commented on src/GafferSceneModule/EditScopeAlgoBinding.cpp:
```diff
@@ -269,6 +269,13 @@ bool renameRenderPassWrapper( Gaffer::EditScope &scope, const std::string &oldNa
+{
+	IECorePython::ScopedGILRelease gilRelease;
+	std::optional<std::string> result = EditScopeAlgo::renameRenderPassNonEditableReason( &scope, newName );
+	return result ? result.value() : "";
```
Although it'd be reasonable to use "" to mean "editable", since we're using `optional` in C++, I think the right return value for the editable case is `None` in Python.

johnhaddon replied:
Alternatively we could drop `optional` on the C++ side and use "" as our editable value on both sides.

johnhaddon commented on python/GafferSceneUI/RenderPassEditor.py:
```diff
@@ -544,6 +548,11 @@ def __renameSelectedRenderPass( self ) :
 		if renderPassName is not None and renderPassName != selectedRenderPasses[0] :
 
+			nonEditableReason = GafferScene.EditScopeAlgo.renameRenderPassNonEditableReason( editScope, renderPassName )
+			if nonEditableReason != "" :
```
This would be `nonEditableReason is not None` if we took my suggestion for the bindings.

johnhaddon reviewed (commented):
Thanks Murray - quick quibble about the Python bindings inline, but otherwise all LGTM.

murraystevenson commented:
> Thanks Murray - quick quibble about the Python bindings inline, but otherwise all LGTM.

Thanks! I've updated the Python bindings to return `None` for the editable case and rebased. Merging.
